### PR TITLE
chore: wrap error objects to include context (#10592)

### PR DESCRIPTION
**Pull request by nsahai8**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: pkg/client/clientset/versioned/typed/application/v1alpha1/fake/fake_appproject.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ func (c *FakeAppProjects) Delete(ctx context.Context, name string, opts v1.Delet
 	_, err := c.Fake.
 		Invokes(testing.NewDeleteActionWithOptions(appprojectsResource, c.ns, name, opts), &v1alpha1.AppProject{})
 
-	return err
+	return fmt.Errorf("error deleting appProject name %w, error: %w",name,err)
```
Comment (crenshaw-dev) [inline pkg/client/clientset/versioned/typed/application/v1alpha1/fake/fake_appproject.go:94]:
```suggestion
	return err
```

Let's revert this one, I think it's a generated file.

File: applicationset/utils/createOrUpdate.go [hunk 2]
```diff
@@ -88,7 +88,7 @@ func CreateOrUpdate(ctx context.Context, c client.Client, obj client.Object, f c
 // mutate wraps a MutateFn and applies validation to its result
 func mutate(f controllerutil.MutateFn, key client.ObjectKey, obj client.Object) error {
 	if err := f(); err != nil {
-		return err
+		return fmt.Errorf("error mutate wraps a MutateFn: %w", err)
```
Comment (crenshaw-dev) [inline applicationset/utils/createOrUpdate.go:91]:
```suggestion
		return err
```

I think in this case it's fine to just return the error.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Thanks @nsahai8! Can you take a look at the errors in the diff? Looks like `fmt` needs to be imported in a few places.

File: pkg/client/clientset/versioned/typed/application/v1alpha1/fake/fake_appproject.go (hunk 1 above)
Comment (nsahai8) [reply pkg/client/clientset/versioned/typed/application/v1alpha1/fake/fake_appproject.go:94]: sure

File: reposerver/repository/lock.go [hunk 3]
```diff
@@ -41,7 +42,10 @@ func (r *repositoryLock) Lock(path string, revision string, allowConcurrent bool
 		if notify {
 			state.cond.Broadcast()
 		}
-		return err
+		if err != nil {
+	        return nil, fmt.Errorf("error in Lock acquiring lock: %w", err)
```
Comment (crenshaw-dev) [inline reposerver/repository/lock.go:46]:
```suggestion
	        return fmt.Errorf("init closer failed: %w", err)
```

I think this line also needs a bit of indentation.  :-)

File: applicationset/services/scm_provider/gitea.go [hunk 4]
```diff
@@ -35,7 +35,7 @@ func NewGiteaProvider(ctx context.Context, owner, token, url string, allBranches
 	}
 	client, err := gitea.NewClient(url, gitea.SetToken(token), gitea.SetHTTPClient(httpClient))
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating a new git client: %w", err)
```
Comment (crenshaw-dev) [inline applicationset/services/scm_provider/gitea.go:38]:
```suggestion
		return nil, fmt.Errorf("error creating a new gitea client: %w", err)
```
Just noticed this is specific to gitea.

File: util/notification/settings/legacy.go [hunk 5]
```diff
@@ -54,11 +54,11 @@ type legacyServicesConfig struct {
 func mergePatch(orig interface{}, other interface{}) error {
 	origData, err := json.Marshal(orig)
 	if err != nil {
-		return err
+		return fmt.Errorf("error in mergePatch: %w", err)
```
Comment (crenshaw-dev) [inline util/notification/settings/legacy.go:57]:
```suggestion
		return fmt.Errorf("error marshaling json for original: %w", err)
```

File: util/notification/settings/legacy.go [hunk 6]
```diff
@@ -54,11 +54,11 @@ type legacyServicesConfig struct {
...
 	}
 	otherData, err := json.Marshal(other)
 	if err != nil {
-		return err
+		return fmt.Errorf("error in mergePatch: %w", err)
```
Comment (crenshaw-dev) [inline util/notification/settings/legacy.go:61]:
```suggestion
		return fmt.Errorf("error marshaling json for patch: %w", err)
```

File: util/notification/settings/legacy.go [hunk 7]
```diff
@@ -67,17 +67,17 @@ func mergePatch(orig interface{}, other interface{}) error {
 
 	mergedData, err := jsonpatch.MergePatch(origData, otherData)
 	if err != nil {
-		return err
+		return fmt.Errorf("error in mergePatch: %w", err)
```
Comment (crenshaw-dev) [inline util/notification/settings/legacy.go:70]:
```suggestion
		return fmt.Errorf("error creating merge patch: %w", err)
```

File: util/notification/settings/legacy.go [hunk 8]
```diff
@@ -163,7 +163,7 @@ func ApplyLegacyConfig(cfg *api.Config, context map[string]string, cm *v1.Config
 		legacyServices := &legacyServicesConfig{}
 		err := yaml.Unmarshal(notifiersData, legacyServices)
 		if err != nil {
-			return err
+			return fmt.Errorf("error in ApplyLegacyConfig: %w", err)
```
Comment (crenshaw-dev) [inline util/notification/settings/legacy.go:166]:
```suggestion
			return fmt.Errorf("error unmarshaling legacy services config: %w", err)
```

File: util/notification/settings/legacy.go [hunk 9]
```diff
@@ -173,11 +173,11 @@ func ApplyLegacyConfig(cfg *api.Config, context map[string]string, cm *v1.Config
 		legacyCfg := &legacyConfig{}
 		err := yaml.Unmarshal([]byte(configData), legacyCfg)
 		if err != nil {
-			return err
+			return fmt.Errorf("error in ApplyLegacyConfig applying config map: %w", err)
```
Comment (crenshaw-dev) [inline util/notification/settings/legacy.go:176]:
```suggestion
			return fmt.Errorf("error in unmarshaling legacy config: %w", err)
```

File: util/settings/accounts.go [hunk 10]
```diff
@@ -105,7 +105,7 @@ func (mgr *SettingsManager) saveAccount(name string, account Account) error {
 func (mgr *SettingsManager) AddAccount(name string, account Account) error {
 	accounts, err := mgr.GetAccounts()
 	if err != nil {
-		return err
+		return fmt.Errorf("error occurred while AddAccount save an account: %w", err)
```
Comment (crenshaw-dev) [inline util/settings/accounts.go:108]:
```suggestion
		return fmt.Errorf("error getting accounts: %w", err)
```

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Mostly tweaks. I generally have the error message focus on the line immediately above - the line that threw the error.

File: applicationset/services/scm_provider/gitea.go [hunk 11]
```diff
@@ -35,7 +35,7 @@ func NewGiteaProvider(ctx context.Context, owner, token, url string, allBranches
 	}
 	client, err := gitea.NewClient(url, gitea.SetToken(token), gitea.SetHTTPClient(httpClient))
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating a new gitea client: %w", err)
```
Comment (nsahai8) [inline applicationset/services/scm_provider/gitea.go:38]:
```suggestion
		return fmt.Errorf("error creating a new gitea client: %w", err)
```

File: applicationset/services/scm_provider/gitea.go [hunk 12]
```diff
@@ -35,7 +35,7 @@ func NewGiteaProvider(ctx context.Context, owner, token, url string, allBranches
 	}
 	client, err := gitea.NewClient(url, gitea.SetToken(token), gitea.SetHTTPClient(httpClient))
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("error creating a new gitea client: %w", err)
```
Comment (nsahai8) [inline applicationset/services/scm_provider/gitea.go:38]:
```suggestion
		return nil, fmt.Errorf("error creating a new gitea client: %w", err)
```

File: reposerver/repository/lock.go [hunk 13]
```diff
@@ -41,7 +42,10 @@ func (r *repositoryLock) Lock(path string, revision string, allowConcurrent bool
 		if notify {
 			state.cond.Broadcast()
 		}
-		return err
+		if err != nil {
+	        return fmt.Errorf("init closer failed: %w", err)
```
Comment (nsahai8) [inline reposerver/repository/lock.go:46]:
```suggestion
	           return fmt.Errorf("init closer failed: %w", err)
```

File: reposerver/repository/lock.go [hunk 14]
```diff
@@ -41,7 +42,10 @@ func (r *repositoryLock) Lock(path string, revision string, allowConcurrent bool
 		if notify {
 			state.cond.Broadcast()
 		}
-		return err
+		if err != nil {
+	           return fmt.Errorf("init closer failed: %w", err)
```
Comment (nsahai8) [inline reposerver/repository/lock.go:46]:
```suggestion
	           	return fmt.Errorf("init closer failed: %w", err)
```

**Comment by crenshaw-dev** [comment]
@nsahai8 do you need help with the DCO check? I think you could rebase interactive on the master branch and squash all the commits into one signed-off commit.

**Pull request closed without merging**